### PR TITLE
jwt_authn: add missing path_match_policy matcher implementation

### PR DESCRIPTION
**Pull request by kozjan**
Commit Message: Added a matcher implementation for JWT. Right now the [documentation](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto.html#config-route-v3-routematch) says that `path_match_policy` may be used, but currently it throws exception.
Additional Description: Previously I implemented it for [rbac only](https://github.com/envoyproxy/envoy/pull/31447). Would be nice if it could be included as a bug fix in the current Envoy version.
Risk Level: Low?
Testing: unit test
Docs Changes: do not know how to handle it if it would be merged before new major release.
Release Notes:
Platform Specific Features:

### CODE REVIEW

**Comment by kozjan** [comment]
@tyxia @yanavlasov 
hey, I'm bumping it to make sure it's not forgotten

File: source/extensions/filters/http/jwt_authn/matcher.cc
Comment (tyxia) [inline source/extensions/filters/http/jwt_authn/matcher.cc:208]: Can `uri_template_matcher_` become nullptr here?

File: source/extensions/filters/http/jwt_authn/matcher.cc
Comment (tyxia) [inline source/extensions/filters/http/jwt_authn/matcher.cc:203]: createPathMatcher returns `absl::StatusOr<PathMatcherSharedPtr>`

It will good to check the return status and fail early to prevent nullptr access.

**Review by tyxia — COMMENTED** [review]
Thanks for contribution!

/wait

File: source/extensions/filters/http/jwt_authn/matcher.cc
Comment (kozjan) [reply source/extensions/filters/http/jwt_authn/matcher.cc]: hey, I made the changes, can you take a look?

File: source/extensions/filters/http/jwt_authn/matcher.cc
Comment (tyxia) [inline source/extensions/filters/http/jwt_authn/matcher.cc:206]: This `EnvoyException` is not tested currently, right? Will you be able to have a quick test cover it?

**Review by tyxia — COMMENTED** [review]
LGTM, modulo one comment.

Thanks!

File: source/extensions/filters/http/jwt_authn/matcher.cc
Comment (kozjan) [reply source/extensions/filters/http/jwt_authn/matcher.cc:206]: Hey, I added a test with wrong configuration to test this status check.
I'm not sure what exactly you mean by also testing the prevention mechanism for nullptr access - it is tested right now, right? Early failure prevents reaching the `match` call on `uri_template_matcher_`.

File: test/extensions/filters/http/jwt_authn/matcher_test.cc
Comment (kozjan) [inline test/extensions/filters/http/jwt_authn/matcher_test.cc:277]: I wanted to check for something that is specifically related to the passed config, but it seems like there are some differences in local test execution vs here in CI - locally some random whitespace is added, so I settled for checking for `INVALID_ARGUMENT` to not impact local tests.

File: source/extensions/filters/http/jwt_authn/matcher.cc
Comment (tyxia) [reply source/extensions/filters/http/jwt_authn/matcher.cc:206]: yea, early failure is what i initially suggested in this comment https://github.com/envoyproxy/envoy/pull/35072#discussion_r1681196576 and what I hope to have test coverage here.

LG, Thanks for adding the test!

**Review by tyxia — DISMISSED** [review]

**Comment by kozjan** [comment]
one more thing, I didn't update the changelog, bc I'm not sure how it works - would it end up in 1.31.1 as a bug fix? in that case do I add it to `changelogs/current.yaml` or this would cause it to end up in next full release 1.32.0?

**Comment by tyxia** [comment]
> one more thing, I didn't update the changelog, bc I'm not sure how it works - would it end up in 1.31.1 as a bug fix? in that case do I add it to `changelogs/current.yaml` or this would cause it to end up in next full release 1.32.0?

 i am not sure why it is categorized as bug fix. It seems to be a new feature to me?  you can add it to `changelogs/current.yaml`.

**Comment by tyxia** [comment]
@kozjan Do you also want to add release note to changelog 

In the same time, pass this to @yanavlasov for final review and merge (also listed as code owner)

File: source/extensions/filters/http/jwt_authn/matcher.cc
Comment (yanavlasov) [inline source/extensions/filters/http/jwt_authn/matcher.cc:224]: To improve const correctness for the state that is shared across multiple threads, please make this `const`
```suggestion
  const Router::PathMatcherSharedPtr uri_template_matcher_;
```

You will then need to create a function for to initialize uri_template_matcher_ so it can be done in the constructor initializer list.

**Review by yanavlasov — CHANGES_REQUESTED** [review]
/wait

File: source/extensions/filters/http/jwt_authn/matcher.cc
Comment (kozjan) [reply source/extensions/filters/http/jwt_authn/matcher.cc]: done

**Comment by kozjan** [comment]
> > one more thing, I didn't update the changelog, bc I'm not sure how it works - would it end up in 1.31.1 as a bug fix? in that case do I add it to `changelogs/current.yaml` or this would cause it to end up in next full release 1.32.0?
> 
> i am not sure why it is categorized as bug fix. It seems to be a new feature to me? you can add it to `changelogs/current.yaml`.

@tyxia I added it in new features in that case.
I thought of it as a bug fix because it aligns the behaviour of jwt and rbac, and right now documentation shows that path_match_policy can be used, while it ends up throwing an error.
Would it be possible to push this change as a patch to v1.30 and v1.31?
The [related change in rbac](https://github.com/envoyproxy/envoy/pull/31447) is present since v1.30.

File: source/extensions/filters/http/jwt_authn/matcher.cc
Comment (yanavlasov) [inline source/extensions/filters/http/jwt_authn/matcher.cc:212]:
```suggestion
  static Router::PathMatcherSharedPtr createUriTemplateMatcher(const RequirementRule& rule) {
```

**Review by yanavlasov — COMMENTED** [review]
Small suggestion to improve initialization safety. LGTM otherwise.

/wait

File: source/extensions/filters/http/jwt_authn/matcher.cc
Comment (kozjan) [reply source/extensions/filters/http/jwt_authn/matcher.cc]: commited your suggestion 👍

Comment (kozjan) [reply source/extensions/filters/http/jwt_authn/matcher.cc]: I think the test that does not pass (`test/integration/quic_http_integration_test`) is unrelated to my changes.

**Comment by yanavlasov** [comment]
LGTM from me. Will wait for final review from @tyxia 

/wait-any

**Comment by kozjan** [comment]
@tyxia, can you take a look?

**Pull request merged**
